README.rst
A Data Algebra® library for database management and computer modeling

.. Algebraix Technology Core Library documentation.
   Copyright Permission.io, Inc. (formerly known as Algebraix Data Corporation), Copyright (c) 2022.

   This file is part of algebraixlib <http://github.com/AlgebraixData/algebraixlib>.

   algebraixlib is free software: you can redistribute it and/or modify it under the terms of
   version 3 of the GNU Lesser General Public License as published by the Free Software Foundation.

   algebraixlib is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
   even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public License along with algebraixlib.
   If not, see <http://www.gnu.org/licenses/>.

   This file is not included via toctree. Mark it as orphan to suppress the warning that it isn't
   included in any toctree.

:orphan:

|logo|_

algebraixlib
============

What Is It?
-----------

``algebraixlib`` is a library that provides constructs and facilities to harness the fundamentals
of data algebra. Data algebra consists of mathematical constructs that can represent all data, no
matter how it is structured, and the operations on that data. With this, all the advantages of a
mathematically rigorous modeling can be unleashed. See also
`A Beginner's Introduction to Data Algebra`_.

Getting Started
---------------

#.  Make sure you have the required versions of Python and Jupyter Notebook installed (see
    `Requirements`_ below).
#.  Install the ``algebraixlib`` library (see `How to Install`_ below).
#.  Download the `examples`_ from our `GitHub`_ repository.
#.  Try the Hello_World.ipynb example first.

(Alternatively, you can also look at a static version of the notebooks in `nbviewer`_; see the
README file in our `examples`_ directory for direct links. For this you don't need to install or
download anything. You can also start with the simpler hello_world.py. However, you'll lose out
on some math and need to read up on it in our documentation at `Read the Docs`_. )

Documentation and Support
-------------------------

*   Find documentation at `Read the Docs`_.
*   Find the ``pip`` installer on `PyPI`_.
*   Find the source code, the bugtracker and contribute on `GitHub`_.
*   Find tutorials and example code in the `examples`_ directory on GitHub.
*   Post questions about algebraixlib on `Stack Overflow`_ using the tag `[algebraixlib]`_.
*   Post questions about the mathematics of data algebra on `math.stackexchange`_ using the tag
    `[data-algebra]`_.
*   Contact us by `email`_.

See also our `GitHub project page`_. In addition, there is a `book`_ about data algebra.


Detailed Instructions
=====================

Requirements
------------

*   `Python`_: Tested with 3.6.1. Likely to run with Python 3.6.x and later. It may run with earlier
    Python 3 versions, but you may run into issues. Does not run with any version of Python before
    Python 3.

    *   For installing and using multiple versions of Python on the same machine, see
        `Official multiple python versions on the same machine? (Stack Overflow)`_,
        `How to install both Python 2.x and Python 3.x in Windows 7 (Stack Overflow)`_ and
        `A Python Launcher For Windows (Python Insider)`_.

*   `Jupyter Notebook`_: Tested with Jupyter 5.2 (used in the Jupyter notebook tutorials and
    examples).

    *   See `Jupyter Installation`_ for instructions how to install the Jupyter notebook
        (``pip install jupyter``).
    *   If you don't want Jupyter in your system environment, you can install it into a virtual
        environment (see `Creation of virtual environments`_).

How to Install
--------------

If you already have Python installed and are familiar with installing packages, you can install
``algebraixlib`` with ``pip``::

> pip install algebraixlib

Additional user permissions may be necessary to complete the installation. In such a situation,
other options include installing the package for a single user (in the user's home directory)::

> pip install algebraixlib --user <username> 

or in a virtual environment (see `Creation of virtual environments`_).

You can also manually download ``algebraixlib`` from `GitHub`_ or `PyPI`_. To install from a
download, unpack it and run the following command from the top-level source directory (the
directory that contains the file setup.py)::

> python setup.py install

(The same considerations about permissions apply.)

Unit Tests
----------

The unit tests require the following libraries to be installed:

*   `nose`_
*   `coverage`_

To execute the unit tests, download the file `runtests.py`_ and the directory `test`_ into any
location on your system, then run `runtests.py`_::

> mkdir algebraixlib-test
> cd algebraixlib-test
> svn export https://github.com/AlgebraixData/algebraixlib/trunk/runtests.py
> svn export https://github.com/AlgebraixData/algebraixlib/trunk/test
> python runtests.py

Documentation Build
-------------------

The documentation build requires the following libraries be installed:

*   `Sphinx`_ (1.3.2 or later)

To run a documentation build, you need a local working copy of our complete `GitHub`_ repository.
Then run `build.py`_ in the directory `docs`_::

> mkdir algebraixlib
> cd algebraixlib
> svn export https://github.com/AlgebraixData/algebraixlib/trunk
> cd trunk/docs
> python build.py


Legalese
========

Copyright
---------

Copyright Permission.io, Inc. (formerly known as Algebraix Data Corporation), Copyright (c) 2022.

License
-------

``algebraixlib`` is free software: you can redistribute it and/or modify it under the terms of
`version 3 of the GNU Lesser General Public License`_ as published by the
`Free Software Foundation`_. A copy of the GNU Lesser General Public License is published along
with ``algebraixlib`` on `GitHub`_. Otherwise, see `GNU licenses`_.

Warranty
--------

``algebraixlib`` is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU Lesser General Public License for more details.

.. _[algebraixlib]:
    http://stackoverflow.com/questions/tagged/algebraixlib
.. _A Beginner's Introduction to Data Algebra:
    http://algebraixlib.readthedocs.org/en/latest/intro.html
.. _A Python Launcher For Windows (Python Insider):
    http://blog.python.org/2011/07/python-launcher-for-windows_11.html
.. _book:
    https://algebraixdata.com/resources/the-algebra-of-data/
.. _build.py:
    https://github.com/AlgebraixData/algebraixlib/blob/master/docs/build.py
.. _coverage:
    https://pypi.python.org/pypi/coverage
.. _Creation of virtual environments:
    https://docs.python.org/3/library/venv.html
.. _[data-algebra]:
    http://math.stackexchange.com/questions/tagged/data-algebra
.. _docs:
    https://github.com/AlgebraixData/algebraixlib/tree/master/docs
.. _Examples:
    https://github.com/AlgebraixData/algebraixlib/tree/master/examples
.. _Free Software Foundation:
    http://www.fsf.org/
.. _How to install both Python 2.x and Python 3.x in Windows 7 (Stack Overflow):
    http://stackoverflow.com/questions/3809314/how-to-install-both-python-2-x-and-python-3-x-in-windows-7
.. _email:
    mailto:algebraixlib@algebraixdata.com
.. _GitHub:
    http://github.com/AlgebraixData/algebraixlib
.. _GitHub project page:
    http://algebraixdata.github.io/algebraixlib/
.. _GNU Licenses:
    http://www.gnu.org/licenses/
.. _Hello_World.ipynb:
    https://github.com/AlgebraixData/algebraixlib/blob/master/examples/Hello_World.ipynb
.. _Jupyter Installation:
    http://jupyter.readthedocs.org/en/latest/install.html
.. _Jupyter Notebook:
    https://jupyter.org/
.. _math.stackexchange:
    http://math.stackexchange.com/
.. _nbviewer:
    http://nbviewer.ipython.org/
.. _nose:
    https://pypi.python.org/pypi/nose/
.. _Official multiple python versions on the same machine? (Stack Overflow):
    http://stackoverflow.com/questions/2547554/official-multiple-python-versions-on-the-same-machine
.. _PyPI:
    http://pypi.python.org/pypi/algebraixlib
.. _Python:
    http://python.org
.. _Read the Docs:
    http://algebraixlib.rtfd.org/
.. _runtests.py:
    https://github.com/AlgebraixData/algebraixlib/blob/master/runtests.py
.. _Sphinx:
    https://pypi.python.org/pypi/Sphinx
.. _Stack Overflow:
    http://stackoverflow.com/
.. _test:
    https://github.com/AlgebraixData/algebraixlib/tree/master/test
.. _Version 3 of the GNU Lesser General Public License:
    http://www.gnu.org/licenses/lgpl-3.0-standalone.html

.. |logo| image:: https://raw.githubusercontent.com/AlgebraixData/algebraixlib/gh-pages/ALGBX-Logo-Color-150DPI.png
.. _logo: http://www.algebraixdata.com/technology/#algebraix-library


RELEASES.rst
.. Algebraix Technology Core Library documentation.
    Copyright Permission.io, Inc. (formerly known as Algebraix Data Corporation), Copyright (c) 2022.

    This file is part of algebraixlib <http://github.com/AlgebraixData/algebraixlib>.

    algebraixlib is free software: you can redistribute it and/or modify it under the terms of
    version 3 of the GNU Lesser General Public License as published by the Free Software Foundation.

    algebraixlib is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
    even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
    Lesser General Public License for more details.

    You should have received a copy of the GNU Lesser General Public License along with algebraixlib.
    If not, see <http://www.gnu.org/licenses/>.

    This file is not included via toctree. Mark it as orphan to suppress the warning that it isn't
    included in any toctree.

:orphan:

Release Notes for algebraixlib
==============================

1.4
---

-   Moved ``CacheStatus`` into its own module and removed ``utils.py`` in ``mathobjects``.
-   Moved all of the modules in ``io`` into ``import_export``.
-   Allow ``import_json`` to take in floats.
-   Better handling of less than functionality for couplets, multisets, and sets.
-   Added a ``tmp_sqlda_op``  to handle nodes that are not "SQL-DA" operations.

1.3
---

-   ``Atom``:

    -   Bug fix to not compare differently typed values as equal.
    -   Updated to have pickle support.

-   mojson.py extended to support ``Multiset``.
-   ``import_csv`` function:

    -   Updated to have new argument for column names.
    -   Updated to propagate multiclan properties.

-   ``Undef``:

    -   Updated to have a factory function.
    -   ``Undef`` propagation added to all algebras to support correct answers for chaining
        data algebra expressions.

-   ``Multiset``:

    -   Iteration performance improved.
    -   Updated to support ``[]`` notation.

-   ``Mathobject`` updated to have ``__deepcopy__`` method.
-   ``unary_multi_extend`` updated to support either ``Set`` or ``Multiset``.
-   ``Couplet`` updated to have factory function.
-   ``Set`` updated to have factory function.
-   Partition functions extended to operate on ``Multiset``\s.
-   ``Multiclan`` bug fix for multiplicities in substrict/superstrict operations.
-   Minor quality of life improvements to ``mo_to_str`` function.
-   ``Multiclan`` algebra updated to have new operations:

    -   ``get_rights``
    -   ``get_rights_for_left``
    -   ``project``
    -   ``from_dict``
    -   ``diag``
    -   ``defined_at``
    -   ``order_slice_to_listgen``
    -   ``order_slice``
    -   ``multiclan_to_listgen``

-   ``Multiset`` algebra updated to have new operations:

    -   ``get_left_set``
    -   ``get_right_set``
    -   ``single``
    -   ``some``

-   Minor quality of life changes to miscellaneous.py.

1.2
---

-   Major improvements to how ``MathObject`` properties are handled.
-   Corrections to propagations of ``MathObject`` properties.
-   Additions to math documentation.
-   Added an example IPython notebooks about XML representation.
-   Minor corrections to example code in IPython notebooks.
-   Minor improvements to layout of documentation.

1.1
---

-   Initial release.


setup_data.py
"""Provide the data for the ``setup()`` call in setup.py.

This is separate here so that other places in our build system that need the algebraixlib data
can get it from here without actually running setup.py.
"""

# Copyright Permission.io, Inc. (formerly known as Algebraix Data Corporation), Copyright (c) 2022.
#
# Confidential and proprietary. For internal use only.
# --------------------------------------------------------------------------------------------------
from setuptools import find_packages

# RELEASE: This is the only place where the algebraixlib version is defined. Other places that
# need the version import this module and get it from here.
version = '1.4'


# Import readme as long description that is used by PyPI as the home page for the library's package.
with open('README.rst') as file:
    long_description = file.read()

# Define the keyword arguments that are sent to setup().
setup_data = {
    'name': 'algebraixlib',
    'version': version,
    'description': 'A data algebra library',
    'long_description': long_description,
    'author': 'Permission.io, Inc.',
    'author_email': 'algebraixlib@algebraixdata.com',
    'maintainer': 'Permission.io, Inc.',
    'maintainer_email': 'algebraixlib@algebraixdata.com',
    'url': 'https://github.com/AlgebraixData/algebraixlib',
    'license': 'http://www.gnu.org/licenses/lgpl-3.0-standalone.html',
    'platforms': [
        'any'
    ],
    'classifiers': [
        'Development Status :: 3 - Alpha',
        'Environment :: Console',
        'Intended Audience :: Developers',
        'Intended Audience :: Education',
        'Intended Audience :: Information Technology',
        'Intended Audience :: Science/Research',
        'License :: OSI Approved :: GNU Lesser General Public License v3 (LGPLv3)',
        'Natural Language :: English',
        'Operating System :: OS Independent',
        'Programming Language :: Python',
        'Programming Language :: Python :: 3',
        'Programming Language :: Python :: 3.4',
        'Programming Language :: Python :: 3.6',
        'Programming Language :: Python :: 3 :: Only',
        'Topic :: Scientific/Engineering',
        'Topic :: Scientific/Engineering :: Information Analysis',
        'Topic :: Scientific/Engineering :: Mathematics',
        'Topic :: Software Development',
        'Topic :: Software Development :: Libraries',
        'Topic :: Software Development :: Libraries :: Python Modules',
    ],
    'packages': find_packages(exclude='test'),
    'install_requires': [
        'rdflib>=4.2'
    ],
    'keywords': 'data algebra set theory',
}
